### PR TITLE
feat(kumactl) upgrade CRDs to apiextensions.k8s.io/v1

### LINKED ISSUE #888: Deprecated `apiextensions.k8s.io/v1beta1` `CustomResourceDefinition`
### Summary

Migrate `CustomResourceDefinition` to `apiextensions.k8s.io/v1` as `apiextensions.k8s.io/v1beta1` is deprecated

```
`CustomResourceDefinition` represents a resource that should be exposed on the API server. Its name MUST
be in the format .spec.name>..spec.group>.
Deprecated in v1.16, planned for removal in v1.19. Use apiextensions.k8s.io/v1 CustomResourceDefinition instead.
```

### Steps To Reproduce

1. Check https://github.com/Kong/kuma/tree/master/pkg/plugins/resources/k8s/native/config/crd/bases

### Additional Details & Logs

- Version
- Error logs
- Configuration
- Platform and Operating System

### PR DESCRIPTION
**Pull request by austince**
### Summary

Upgrades the CRDs from `apiextensions.k8s.io/v1beta1` to `apiextensions.k8s.io/v1`. Installed this on a v1.18 `kind` cluster without issue. 

This will drop support for k8s < 1.16.0, is that ok? 

### Full changelog

* Set `spec.scope` to `Namespaced` if empty
* Move `spec.validation.openAPIV3Schema` to `spec.versions[*].schema.openAPIV3Schema`
  * Remove `metadata` fields from schema validation spec, as they are now implied
* Check for usage of any other deprecated/ changed features here:
  * see: https://github.com/kubernetes/kubernetes/pull/79604

### Issues resolved

Fix #888

### Documentation

This will drop support for k8s < 1.16.0, not sure where this should be documented but happy to make an appropriate PR wherever!

### CODE REVIEW

File: app/kumactl/data/install/k8s/control-plane/crds/kuma.io_circuitbreaker.yaml [hunk 1]
```diff
@@ -10,385 +10,386 @@ spec:
...
-            of an object. Servers should convert recognized schemas to the latest
-            internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#resources'
-          type: string
-        kind:
-          description: 'Kind is a string value representing the REST resource this
-            object represents. Servers may infer this from the endpoint the client
-            submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds'
-          type: string
-        metadata:
+  scope: Namespaced
```
Comment (nickolaev) [inline app/kumactl/data/install/k8s/control-plane/crds/kuma.io_circuitbreaker.yaml:13]: Please keep these as `scope: Cluster`. All our resources are not namespace bound.

Comment (austince) [reply app/kumactl/data/install/k8s/control-plane/crds/kuma.io_circuitbreaker.yaml:13]: Hey, thanks for the review @nickolaev -- these are set to empty strings in the current version, which I thought defaults to "Namespaced". Am I missing something?

Comment (austince) [reply app/kumactl/data/install/k8s/control-plane/crds/kuma.io_circuitbreaker.yaml:13]: The only one that is explicitly set to `Cluster` is `Mesh`.

Refs:
https://github.com/Kong/kuma/blob/f46bc59eeb58f9f7d5295a7c2135aee31d7def0e/app/kumactl/data/install/k8s/control-plane/crds/kuma.io_circuitbreaker.yaml#L10-L13

https://github.com/Kong/kuma/blob/f46bc59eeb58f9f7d5295a7c2135aee31d7def0e/app/kumactl/data/install/k8s/control-plane/crds/kuma.io_meshes.yaml#L10-L13

Comment (nickolaev) [reply app/kumactl/data/install/k8s/control-plane/crds/kuma.io_circuitbreaker.yaml:13]: Yes, that is true. On the other hand, we only access these from `kuma-system` (i.e. where the kuma-cp is deployed) so maybe `Namespaced` is good enough. Let's go with it.

**Comment by austince** [comment]
Hey @nickolaev, got the tests updated and code regenerated. Still have an issue w/ [some bits of the schema migration](https://app.circleci.com/pipelines/github/Kong/kuma/3729/workflows/930311aa-4057-4a9c-852f-0b52e524cdef/jobs/51306), but I'm also noticing v1.13 [fails because the API doesn't exist](https://app.circleci.com/pipelines/github/Kong/kuma/3729/workflows/930311aa-4057-4a9c-852f-0b52e524cdef/jobs/51305). Before I fix the schema stuff, just want to check with you that this is still something we should merge, or if there's anything else I should do instead. Thank you!

**Comment by nickolaev** [comment]
OK here is my plan:
 * please, do keep moving this PR forward
 * we might keep it around until 1.19 is out (https://github.com/kubernetes/sig-release/tree/master/releases/release-1.19)
 * after that we'll have to either drop support for <1.16.x and support >1.19.x or stick with the old API for another couple of months.

In any case - I see this as a valuable PR that is tackling the API migration early so we are prepared.


It might be useful to find what are the K8s versions used in the major public clouds so we are aware if we drop `1.13` are we dropping our users' use-cases.

**Comment by austince** [comment]
Sounds like a good plan to me! I'll make those changes either today or tomorrow, as well as put together a list of current cloud provider versions.

**Comment by austince** [comment]
Current supports k8s versions:

GKE: [as of Jun 29](https://cloud.google.com/kubernetes-engine/docs/release-notes#june_29_2020)
* lowest: 1.14, though seems to be retiring
* highest: 1.17

AKE: [as of Jun 29](https://github.com/Azure/AKS/releases/tag/2020-06-29)
* lowest: 1.14, retiring
* highest: 1.17

EKS: [as of Jul 10](https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html)
* lowest: 1.14
* highest: 1.16

Tencent TKE: a bit difficult to tell: https://intl.cloud.tencent.com/document/product/457/9315
* lowest: 1.7 (?)
* highest: 1.16

OpenShift: [as of 4.3](https://docs.openshift.com/container-platform/4.3/release_notes/ocp-4-3-release-notes.html) 
* uses version: 1.16
* last supported version, w/ openshift 4.2, was 1.14


Alibaba Cloud: [as of Jun 28](https://www.alibabacloud.com/help/doc-detail/98310.htm?spm=a2c63.l28256.a3.20.20d37d1bKYiAPG)
* [official policy](https://www.alibabacloud.com/help/doc-detail/115453.htm?spm=a2c63.l28256.b99.40.20d37d1bKYiAPG) states support for last two major release
* lowest: 1.12 (only for certain users), and 1.14 for GA
* highest: 1.16


Since 1.16 is the highest by most, I'd agree with you that we should wait a few more months but good to get ahead of it.

The current build seems to be failing because [the `trafficpermission/everything` resource seems to be invalid](https://app.circleci.com/pipelines/github/Kong/kuma/3740/workflows/176db9a4-7efc-47af-9dd2-3b1050d88343/jobs/51514). Do you know if this is caused by these changes, or is there another place I should look to update that?

**Comment by nickolaev** [comment]
Installing Kuma 0.7.1 on K8s 1.19.2 with a couple of warnings:
```shell
$ ./kumactl install control-plane | kubectl apply -f -
namespace/kuma-system created
secret/kuma-sds-tls-cert created
secret/kuma-kds-tls-cert created
secret/kuma-admission-server-tls-cert created
configmap/kuma-control-plane-config created
serviceaccount/kuma-control-plane created
Warning: apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
customresourcedefinition.apiextensions.k8s.io/circuitbreakers.kuma.io created
customresourcedefinition.apiextensions.k8s.io/dataplaneinsights.kuma.io created
customresourcedefinition.apiextensions.k8s.io/traffictraces.kuma.io created
customresourcedefinition.apiextensions.k8s.io/zones.kuma.io created
customresourcedefinition.apiextensions.k8s.io/zoneinsights.kuma.io created
customresourcedefinition.apiextensions.k8s.io/dataplanes.kuma.io created
customresourcedefinition.apiextensions.k8s.io/faultinjections.kuma.io created
customresourcedefinition.apiextensions.k8s.io/healthchecks.kuma.io created
customresourcedefinition.apiextensions.k8s.io/meshes.kuma.io created
customresourcedefinition.apiextensions.k8s.io/proxytemplates.kuma.io created
customresourcedefinition.apiextensions.k8s.io/trafficlogs.kuma.io created
customresourcedefinition.apiextensions.k8s.io/trafficpermissions.kuma.io created
customresourcedefinition.apiextensions.k8s.io/trafficroutes.kuma.io created
clusterrole.rbac.authorization.k8s.io/kuma:control-plane created
clusterrolebinding.rbac.authorization.k8s.io/kuma:control-plane created
role.rbac.authorization.k8s.io/kuma:control-plane created
rolebinding.rbac.authorization.k8s.io/kuma:control-plane created
service/kuma-control-plane created
deployment.apps/kuma-control-plane created
Warning: admissionregistration.k8s.io/v1beta1 MutatingWebhookConfiguration is deprecated in v1.16+, unavailable in v1.22+; use admissionregistration.k8s.io/v1 MutatingWebhookConfiguration
mutatingwebhookconfiguration.admissionregistration.k8s.io/kuma-admission-mutating-webhook-configuration created
Warning: admissionregistration.k8s.io/v1beta1 ValidatingWebhookConfiguration is deprecated in v1.16+, unavailable in v1.22+; use admissionregistration.k8s.io/v1 ValidatingWebhookConfiguration
validatingwebhookconfiguration.admissionregistration.k8s.io/kuma-validating-webhook-configuration created
```

**Comment by nickolaev** [comment]
@austince would you consider porting these changes to the Helm charts? I think we should consider these updates after 1.0.0 is released.

**Comment by austince** [comment]
Will start reworking this at the beginning of next week -- the only issue I think we might run into is generating the `spec` fields, if necessary. Kubebuilder has a way to do this, I may try to leverage that.

**Comment by austince** [comment]
Closed by #1108

**Pull request closed without merging**